qubekit/utils/datastructures.py

```diff
@@ -211,6 +211,16 @@ def is_available(cls) -> bool:
         """Check any dependencies to make sure that this stage is available to run."""
         ...
 
+    def _run(self, molecule: "Ligand", *args, **kwargs) -> "Ligand":
+        """run it on the molecule and on the fragments """
+        molecule = self.run(molecule, *args, **kwargs)
+
+        # run it on the fragments
+        if molecule.fragments is not None:
+            molecule.fragments = [self.run(fragment, *args, fragment=True, **kwargs) for fragment in molecule.fragments]
+
+        return molecule
+
```

Comment:
Can we swap this round and have this be run and this call the abstract method `_run` which is to be created for each subclass and acts on each ligand separately? Then if we use any stages outside of the main workflow the public `run` will run the parent and fragments by default rather than having to call `_run`. This way the main `run` method can also handle any special actions which depend on the parent and the fragments at the same time, the main on that comes to mind is the torsion parameter optimisation.

Reply:
Sure, we could use a more meaningful name for the coordinator like `run` and `run_coordinator` or `run_single`